jxl/src/headers/permutation.rs

```diff
@@ -57,23 +57,168 @@ impl Permutation {
             prev_val = val;
         }
 
-        let mut temp = Vec::new();
-        temp.try_reserve((size - skip) as usize)?;
-        temp.extend((skip as usize)..(size as usize));
+        // Create a temporary permutation vector for the elements to permute
+        let mut perm_temp: Vec<u32> = (skip..size).collect();
 
-        let mut permutation = Vec::new();
-        permutation.try_reserve(size as usize)?;
+        // Decode the Lehmer code
+        decode_lehmer_code(&lehmer, &mut perm_temp)?;
 
-        permutation.extend(0..skip as usize);
-        for idx in lehmer {
-            permutation.push(temp.remove(idx as usize));
-        }
-        permutation.extend(temp);
+        // Construct the full permutation vector
+        let mut permutation = Vec::with_capacity(size as usize);
+        permutation.extend(0..(skip as usize)); // Add skipped elements
+
+        // Append the permuted elements
+        permutation.extend(perm_temp.iter().map(|&x| x as usize));
+
+        // Ensure the permutation has the correct size
+        assert_eq!(permutation.len(), size as usize);
 
         Ok(Self(permutation))
     }
 }
 
+// Returns the value of the lowest set bit.
+fn value_of_lowest_1_bit(t: u32) -> u32 {
+    t & t.wrapping_neg()
+}
+
+// Returns base-2 logarithm, rounded down.
+fn floor_log2_nonzero(x: u32) -> usize {
+    31 - x.leading_zeros() as usize
+}
+
+// Returns base-2 logarithm, rounded up.
+fn ceil_log2_nonzero(x: u32) -> usize {
+    let floor_log2 = floor_log2_nonzero(x);
+    if (x & (x - 1)) == 0 {
+        // power of two
+        floor_log2
+    } else {
+        floor_log2 + 1
+    }
+}
+
+// Decodes the Lehmer code in code[0..n) into permutation[0..n).
+fn decode_lehmer_code(code: &[u32], permutation: &mut [u32]) -> Result<()> {
+    println!("code: {:?}", code);
+    println!("permutation: {:?}", permutation);
```

Comment:
`println!` could be replaced with `#[tracing::instrument]` attribute on the function.

Reply:
I just removed those left over `println`s